src/test/java/com/thomashan/collection/immutable/ImmutableListImplTest.java

```diff
@@ -0,0 +1,253 @@
+        Object object = new Object();
+
+        assertEquals(0, of(object, object).indexOf(object));
+    }
+
+    @Test
+    public void testIndexOf_GivenObjectIsNotInTheList_ReturnsNegativeOne() {
+        assertEquals(-1, of().indexOf(new Object()));
+    }
+
+    @Test
+    public void testLastIndexOf_GivenObjectIsInTheList_ReturnsLastIndex() {
+        Object object = new Object();
+
+        assertEquals(1, of(object, object).lastIndexOf(object));
+    }
+
+    @Test
+    public void testLastIndexOf_GivenObjectIsNotInTheList_ReturnsNegativeOne() {
+        assertEquals(-1, of().lastIndexOf(new Object()));
+    }
+
+    @Test
+    public void testListIterator_ReturnsListIterator() {
+        assertTrue(of().listIterator() instanceof ListIterator);
+    }
+
+    @Test
+    public void testListIteratorWithIndex_ReturnsListIterator() {
+        assertTrue(of().listIterator(0) instanceof ListIterator);
+    }
+
+    @Test
+    public void testListIteratorWithIndex_ReturnsListIteratorWithCorrectPosition() {
+        Object object = new Object();
+
+        assertEquals(object, of(new Object(), object).listIterator(1).next());
+    }
+
+    @Test
+    public void testSubList_ReturnsImmutableList() {
+        Object object = new Object();
+
+        assertTrue(of(object, object).subList(0, 0) instanceof ImmutableList);
+    }
+
+    @Test
+    public void testSubList_ReturnImmutableListWithRightElements() {
+        Object object = new Object();
+
+        assertEquals(of(object), of(object, object).subList(0, 1));
+    }
+
+    @Test
+    public void testPlus_ReturnsImmutableList() {
+        assertTrue(of().plus(new Object()) instanceof ImmutableList);
+    }
+
+    @Test
+    public void testPlus_ReturnsNewInstanceOfImmutableList() {
+        ImmutableList<Object> list = of();
+
+        assertFalse(list == list.plus(new Object()));
+    }
+
+    @Test
+    public void testPlus_ReturnsNewInstanceOfImmutableListWithNewElementAdded() {
+        assertEquals(1, of().plus(new Object()).size());
+    }
+
+    @Test
+    public void testMinus_ReturnsImmutableList() {
+        assertTrue(of().minus(new Object()) instanceof ImmutableList);
+    }
+
+    @Test
+    public void testMinus_ReturnsNewInstanceOfImmutableList() {
+        Object object = new Object();
+        ImmutableList<Object> list = of(object);
+
+        assertFalse(list == list.minus(object));
+    }
+
+    @Test
+    public void testMinus_ReturnsNewInstanceOfImmutableListWithNewElementAdded() {
+        Object object = new Object();
+
+        assertEquals(0, of().minus(object).size());
+    }
+
+    @Test
+    public void testAddOrSet_ReturnsImmutableList() {
+        assertTrue(of().addOrSet(0, new Object()) instanceof ImmutableList);
+    }
+
+    @Test
+    public void testAddOrSet_ReturnsNewInstanceOfImmutableList() {
+        Object object = new Object();
+        ImmutableList<Object> list = of(object);
+
+        assertFalse(list == list.addOrSet(0, object));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use assertSame(x, y) instead of assertTrue(x==y), or assertNotSame(x,y) vs assertFalse(x==y)](https://app.codacy.com/app/thomashan/coup-java/pullRequest?prid=4509131)